src/transform/src/literal_constraints.rs

```diff
@@ -0,0 +1,637 @@
+        let mut should_inline = vec![false; mfp.input_arity + mfp.expressions.len()];
+        // Mark those expressions for inlining that contain a subexpression of the form
+        // `<xxx> = <lit>` or `<lit> = <xxx>`.
+        for (i, e) in mfp.expressions.iter().enumerate() {
+            e.visit_post(&mut |s| {
+                if s.any_expr_eq_literal().is_some() {
+                    should_inline[i + mfp.input_arity] = true;
+                }
+            })?;
+        }
+        // Whenever
+        // `<Column(i)> = <lit>` or `<lit> = <Column(i)>`
+        // appears in a predicate, mark the ith expression to be inlined.
+        for (_before, p) in mfp.predicates.iter() {
+            p.visit_post(&mut |e| {
+                if let MirScalarExpr::CallBinary {
+                    func: BinaryFunc::Eq,
+                    expr1,
+                    expr2,
+                } = e
+                {
+                    if matches!(**expr1, MirScalarExpr::Literal(..)) {
+                        if let MirScalarExpr::Column(col) = **expr2 {
+                            if col >= mfp.input_arity {
+                                should_inline[col] = true;
+                            }
+                        }
+                    }
+                    if matches!(**expr2, MirScalarExpr::Literal(..)) {
+                        if let MirScalarExpr::Column(col) = **expr1 {
+                            if col >= mfp.input_arity {
+                                should_inline[col] = true;
+                            }
+                        }
+                    }
+                }
+            })?;
+        }
+        // Perform the marked inlinings.
+        mfp.perform_inlining(should_inline);
+        Ok(())
+    }
+
+    /// MFPs have a Vec of predicates [p1, p2, ...], which logically represents `p1 AND p2 AND ...`.
+    /// This function performs this conversion. Note that it might create a variadic AND with
+    /// 0 or 1 args, so the resulting predicate Vec always has exactly 1 element.
+    fn list_of_predicates_to_and_of_predicates(mfp: &mut MapFilterProject) {
+        // Rebuild the MFP. (Unfortunately, we cannot modify the predicates in place, because MFP
+        // predicates also have a "before" field, which we need to update. (`filter` will recompute
+        // these.)
+        let (map, _predicates, project) = mfp.as_map_filter_project();
+        let new_predicates = vec![MirScalarExpr::CallVariadic {
+            func: VariadicFunc::And,
+            exprs: mfp.predicates.iter().map(|(_, p)| p.clone()).collect(),
+        }];
+        *mfp = MapFilterProject::new(mfp.input_arity)
+            .map(map)
+            .filter(new_predicates)
+            .project(project);
+    }
+
+    /// Call [mz_expr::canonicalize::canonicalize_predicates] on each of the predicates in the MFP.
+    fn canonicalize_predicates(mfp: &mut MapFilterProject, relation: &MirRelationExpr) {
```

Comment:
I'm reading through this, and the main confusion I have is what the bar is for inclusion in `literal_constraints.rs`. I did not read deeply, but superficially this name makes it sound like it it should be in `canonicalize_mfps.rs`? There are several "expression mechanics" in this file, and might (in the future, now, somewhere in between) they find a more specific home?

Reply:
The main call to `canonicalize_predicates` is in `fusion/filter.rs`. Now that you mention it, this indeed seems a bit strange. It's also strange that `CanonicalizeMfp` also does the actual fusion part of filter fusion. I created an issue to sort this out: "Merge filter fusion and `CanonicalizeMfp`" https://github.com/MaterializeInc/materialize/issues/17142

Reply:
And the rest of the expression manipulation in `literal_constraints.rs` is strictly to support the literal constraints detection. (Btw. I have an issue to refactor literal constraints detection into a recursive implementation from the current disjunctive normal form-based implementation: https://github.com/MaterializeInc/materialize/issues/14797)